geometric/internal.py

```diff
@@ -3483,4 +3500,443 @@ def __init__(self, molecule, **kwargs):
+                self.ImageICs.append(PrimitiveInternalCoordinates(molecule[i], connect=connect, addcart=addcart, **kwargs))
+                self.CartICs.append(CartesianCoordinates(molecule[i], **kwargs))
+        self.ImageICs.append(EmptyCoordinates(molecule[-1], **kwargs))
+        self.CartICs.append(EmptyCoordinates(molecule[-1], **kwargs))
+        for i, imageIC in self.ICIter():
+            for ic in imageIC.Internals:
+                self.Internals.append(ImagePrim(ic, self.na, i))
+        guessw = kwargs.get('guessw', 0.1)
+        # kwargs['cartesian'] = True
+        if kwargs.get('cartesian', True):
+            #print("cartesian, w =", guessw)
+            for i, imageIC in self.ICIter():
+                self.Internals.append(RMSDisplacement(imageIC, self.na, i-1, i, head=molecule.xyzs[0]*ang2bohr, tail=molecule.xyzs[-1]*ang2bohr, w=guessw))
+            self.Internals.append(RMSDisplacement(imageIC, self.na, i+1, i, head=molecule.xyzs[0]*ang2bohr, tail=molecule.xyzs[-1]*ang2bohr, w=guessw))
+        else:
+            if self.ic_displace:
+                print("icdisp, w =", guessw)
+                for i, imageIC in self.ICIter():
+                    self.Internals.append(RMSDisplacement(imageIC, self.na, i-1, i, head=molecule.xyzs[0]*ang2bohr, tail=molecule.xyzs[-1]*ang2bohr, w=guessw))
+                self.Internals.append(RMSDisplacement(imageIC, self.na, i+1, i, head=molecule.xyzs[0]*ang2bohr, tail=molecule.xyzs[-1]*ang2bohr, w=guessw))
+            else:
+                print("ccdisp, w =", guessw)
+                for i, cartIC in self.CCIter():
+                    self.Internals.append(RMSDisplacement(cartIC, self.na, i-1, i, head=molecule.xyzs[0]*ang2bohr, tail=molecule.xyzs[-1]*ang2bohr, w=guessw))
+                self.Internals.append(RMSDisplacement(cartIC, self.na, i+1, i, head=molecule.xyzs[0]*ang2bohr, tail=molecule.xyzs[-1]*ang2bohr, w=guessw))
+                
+        # for i, imageIC in self.ICIter():
+        #     self.Internals.append(ChainKink(imageIC, self.na, i-1, i+1, i, head=molecule.xyzs[0]*ang2bohr, tail=molecule.xyzs[-1]*ang2bohr))
+
+    def __repr__(self):
+        lines = ["Internal coordinate system (atoms numbered from 1):"]
+        typedict = OrderedDict()
+        for Internal in self.Internals:
+            lines.append(Internal.__repr__())
+            if str(Internal.get_type()) not in typedict:
+                typedict[str(Internal.get_type())] = 1
+            else:
+                typedict[str(Internal.get_type())] += 1
+        if len(lines) > 100:
+            # Print only summary if too many
+            lines = []
+        for k, v in list(typedict.items()):
+            lines.append("%s : %i" % (k, v))
+        return '\n'.join(lines)
+
+    def clearCache(self):
+        super(ChainCoordinates, self).clearCache()
+        for i, c in self.ICIter():
+            c.clearCache()
+        for i, c in self.CCIter():
+            c.clearCache()
+    
+    def ICIter(self):
```

Comment:
I wonder why the loop does not include internal coordinate systems of the endpoints?